pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -124,7 +124,19 @@ public boolean isDuplicate(MessageId messageId) {
             // Already included in a cumulative ack
             return true;
         } else {
-            return pendingIndividualAcks.contains(MessageIdAdvUtils.discardBatch(messageIdAdv));
+            // If "batchIndexAckEnabled" is false, the batched messages acknowledgment will be traced by
+            // pendingIndividualAcks. So no matter what type the message ID is, check with "pendingIndividualAcks"
+            // first.
+            if (pendingIndividualAcks.contains(MessageIdAdvUtils.discardBatch(messageIdAdv))) {
+                return true;
+            }
+            if (messageId instanceof BatchMessageIdImpl) {
+                BatchMessageIdImpl batchMessageId = (BatchMessageIdImpl) messageId;
```

Comment:
The following implementation would be more elegant and simple.

```java
            final MessageIdAdv key = MessageIdAdvUtils.discardBatch(messageIdAdv);
            if (pendingIndividualAcks.contains(key)) {
                return true;
            }
            if (messageIdAdv.getBatchIndex() >= 0) {
                ConcurrentBitSetRecyclable bitSet = pendingIndividualBatchIndexAcks.get(key);
                return bitSet != null && !bitSet.get(messageIdAdv.getBatchIndex());
            }
```

1. You can get the batch index from the `messageIdAdv` before.
2. Avoid duplicated calls of `discardBatch`.

Reply:
fixed